### ISSUE TITLE
Program interrupted. This could be caused by a memory limit or an error in your code.

### ISSUE THREAD

**Issue by JhunDigal** [thebaselab/codeapp#818]
Wasn’t able to use this app at all even with a very basic code!

**Comment by MarZim81** [comment]
Same issue here, can‘t execute the „Creating a Node.js“ sample project.

**Comment by bummoblizard** [comment]
Let me know your device model and the iOS version for me to reproduce this problem.

**Comment by MarZim81** [comment]
Sorry, totally forgot.
Device: iPad Pro M2 Gen 6
iOS: 16.4.1

Same issue on my iPhone 13 with same iOS version.

**Comment by hezhenwen502** [comment]
I also encountered this situation, and when using the npm i installation package, this error was reported.
Device: iPad mini6
IOS：15.4.1

**Comment by hezhenwen502** [comment]
> 让我知道您的设备型号和iOS版本，以便我重现此问题。

I also encountered this situation, and when using the npm i installation package, this error was reported.
Device: iPad mini6
IOS：15.4.1

**Comment by sadxsmile** [comment]
same error here

**Comment by AnyUserNameNotAvailable** [comment]
same error here

**Comment by zaychika** [comment]
Same issue.  I think it just comes down to a limitation of this app for error handling.  

TLDR: this error is telling you there's a problem with your code, but the app cannot tell you what that problem is or what file and line its coming from. 

Here's some examples in Node.js:

1.  Running `nodemon app.js`
ERROR: `Program interrupted. This could be caused by a memory limit or an error in your code`
Solution: Just use `node app.js`.  Installing nodemon globally or as an app dependency doesn't work.  Just can't use nodemon with this platform yet.

2. Running `node app.js` with node package "Mongoose"
ERROR: `Program interrupted. This could be caused by a memory limit or an error in your code`
Solution: In my case, I was using deprecated syntax from an old version of mongoose.  Just had to guess that was the problem and update per the docs.

3. Stopping a node app with `ctrl-c`
ERROR: `Program interrupted. This could be caused by a memory limit or an error in your code`
Solution: This is weird.  There really is no problem.  I'm just stopping the process like normal.  But this app handles it like something went wrong.  Kinda misleading.

Takeaway: In each case, there was a simple problem that most other text editor/IDE systems would spell out for you.  In my case, it's almost always an actual problem with the code.  But the app won't declare the problem or where it occurs in the console.  It will list problems there, to be clear.  Just not the ones that cause this particular error message.

Device: iPad Pro 3rd Gen
OS: iPadOS 16.5

**Comment by RenzoAlessandro** [comment]
Same error here… doesn't support creating new angular and react projects?

**Comment by Envo** [comment]
I tried it with a very-very simple react project and i got the same. Something is broken in the application.

> Device: iPad Pro 3rd Gen
OS: iPadOS 17.0 (Beta 4)
Code App Version: 1.4.5

**Comment by bummoblizard** [comment]
> I tried it with a very-very simple react project and i got the same. Something is broken in the application.
> 
> 
> 
> > Device: iPad Pro 3rd Gen
> 
> OS: iPadOS 17.0 (Beta 4)
> 
> Code App Version: 1.4.5

Please use our React sample as a starting point: https://github.com/thebaselab/react-starter.

**Comment by Envo** [comment]
Yeah that's run, but what's the point if my own apps isn't...

> > I tried it with a very-very simple react project and i got the same. Something is broken in the application.
> > > Device: iPad Pro 3rd Gen
> > 
> > 
> > OS: iPadOS 17.0 (Beta 4)
> > Code App Version: 1.4.5
> 
> Please use our React sample as a starting point: https://github.com/thebaselab/react-starter.

Yeah that's run without an issue, but what's the point of an IDE if It can't start my own projects...

**Comment by bummoblizard** [comment]
> > > I tried it with a very-very simple react project and i got the same. Something is broken in the application.
> 
> > > > Device: iPad Pro 3rd Gen
> 
> > > 
> 
> > > 
> 
> > > OS: iPadOS 17.0 (Beta 4)
> 
> > > Code App Version: 1.4.5
> 
> > 
> 
> > Please use our React sample as a starting point: https://github.com/thebaselab/react-starter.
> 
> 
> 
> Yeah that's run, but what's the point if my own projects isn't... 

You can. It's just certain commands aren't available on Code App because of iOS's restriction. Usually, they try to spawn sub-processes.

You might send me a copy of your project at sh.chung@thebaselab.com and I will patch it for you.

**Comment by Envo** [comment]
> > > > I tried it with a very-very simple react project and i got the same. Something is broken in the application.
> > 
> > 
> > > > > Device: iPad Pro 3rd Gen
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > OS: iPadOS 17.0 (Beta 4)
> > 
> > 
> > > > Code App Version: 1.4.5
> > 
> > 
> > > 
> > 
> > 
> > > Please use our React sample as a starting point: https://github.com/thebaselab/react-starter.
> > 
> > 
> > Yeah that's run, but what's the point if my own projects isn't...
> 
> You can. It's just certain commands aren't available on Code App because of iOS's restriction. Usually, they try to spawn sub-processes.
> 
> You might send me a copy of your project at [sh.chung@thebaselab.com](mailto:sh.chung@thebaselab.com) and I will patch it for you.

I take your answer as a kind joke. :) Of course i won't going to share these projects mostly because of security concerns...

**Comment by Flmob** [comment]
> > > > I tried it with a very-very simple react project and i got the same. Something is broken in the application.
> > 
> > 
> > > > > Device: iPad Pro 3rd Gen
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > OS: iPadOS 17.0 (Beta 4)
> > 
> > 
> > > > Code App Version: 1.4.5
> > 
> > 
> > > 
> > 
> > 
> > > Please use our React sample as a starting point: https://github.com/thebaselab/react-starter.
> > 
> > 
> > Yeah that's run, but what's the point if my own projects isn't...
> 
> You can. It's just certain commands aren't available on Code App because of iOS's restriction. Usually, they try to spawn sub-processes.
> 
> You might send me a copy of your project at [sh.chung@thebaselab.com](mailto:sh.chung@thebaselab.com) and I will patch it for you.


Can you prrovide more info in wiki about starting node project in app? About app limitations, possible errors and how to start hevier that small epress eample apps. I believe this will help to prevent a lot of possible questions in the future

**Comment by adamszigeti** [comment]
I seem to have found a solution by wrapping my code in a try/catch block, and just console.logging/console.error-ing the error out (`process.on("unhandledException", ...)` and `process.on("unhandledRejection", ...)` did not work).

Though that alone won't work with async code. Instead, I suggest you wrap your whole program in a function, export it, and then use a new entry file like so:

```
// bootstrap.js
try {
  const { myWrappedProgram } = require("./myprogram.js")
  const result = myWrappedProgram()
  
  if (result && typeof result === "object" && "then" in result) {
    result.then(() => {}, console.error)
  }
}
catch (error) {
  console.error(error)
}

// myprogram.js
module.exports = { myWrappedProgram }

async function myWrappedProgram() {
  // ...
}

// to automatically start when the script is directly run via "node myprogram.js":
if (typeof require !== 'undefined' && require.main === module) {
    myWrappedProgram();
}
```

Now you start like so: `node bootstrap.js`.

I think it's actually really silly we have to do this, but until a fix is avaiable, this is better than nothing.

**Comment by bummoblizard** [comment]
Thanks all for the discussion. While it is not perfect, https://github.com/thebaselab/codeapp/commit/b0c6519b2340743b4afa7f8a968008c1c7eef39e should make uncaught errors reported properly. Additionally, node version is now updated to 18.

<img width="1080" alt="image" src="https://github.com/thebaselab/codeapp/assets/38398443/8104cbc8-ed29-480d-aad9-307d1c1ae98f">
